dev-tool/src/list_net/inspect.rs

```diff
@@ -0,0 +1,662 @@
+    //   automatically select the right deserializer and prevent confusion when
+    //   accidentally mixing up this files.
+    /// Type of the file (either "binparams" or "samples")
+    #[structopt(short, long, parse(try_from_str))]
+    r#type: FileType,
+
+    /// Prints the elements of all inspected matrices.
+    #[structopt(short, long)]
+    print_data: bool,
+
+    /// If set only skip arrays which names are not in the filter.
+    ///
+    /// This accepts comma separated lists, as well as including the option multiple times.
+    #[structopt(short, long)]
+    filter: Option<Vec<String>>,
+
+    /// Prints stats including `min`, `max`, `mean`, `std`, `has_nans`, `has_infs`, `has_subnormals`.
+    #[structopt(short = "s", long)]
+    stats: bool,
+
+    /// Checks if all elements in all matrices are "normal".
+    #[structopt(short = "c", long)]
+    check_normal: bool,
+
+    /// Checks if all values in all matrices are in given range.
+    ///
+    /// Format: <from>..=<to>
+    ///
+    /// E.g.: --check-range="-10..=20"
+    ///
+    /// Currently only full inclusive ranges are supported.
+    #[structopt(short = "r", long)]
+    check_range: Option<String>,
+
+    /// Path to a input file.
+    file: PathBuf,
+}
+
+#[derive(Debug, Display, Clone, Copy)]
+/// Supported file types.
+pub enum FileType {
+    /// BinParams file
+    BinParams,
+    /// Samples file
+    Samples,
+}
+
+impl FromStr for FileType {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use FileType::*;
+        match &*s.trim().to_lowercase() {
+            "binparams" | "bin-params" => Ok(BinParams),
+            "samples" => Ok(Samples),
+            _ => bail!("Unexpected file type. Supported types are: \"binparams\", \"samples\""),
+        }
+    }
+}
+
+type MatricesIter = Box<dyn Iterator<Item = Result<(String, ArrayD<f32>), Error>>>;
+
+impl InspectCmd {
+    pub fn run(self) -> Result<i32, Error> {
+        debug!("Loading Matrices from {}.", self.r#type);
+        let matrices = self.load_matrices()?;
+
+        debug!("Inspecting Matrices");
+        self.run_inspect_matrices(matrices)
+    }
+
+    fn load_matrices(&self) -> Result<MatricesIter, Error> {
+        match self.r#type {
+            FileType::BinParams => {
+                let bin_params = BinParams::deserialize_from_file(&self.file)?;
+                Ok(bin_params_to_matrices_iter(bin_params))
+            }
+            FileType::Samples => {
+                let samples = InMemorySamples::deserialize_from_file(&self.file)?;
+                samples_to_matrices_iter(samples)
+            }
+        }
+    }
+
+    fn run_inspect_matrices(self, matrices: MatricesIter) -> Result<i32, Error> {
+        let Self {
+            r#type: _,
+            print_data,
+            file: _,
+            stats,
+            check_normal,
+            check_range,
+            filter,
+        } = self;
```

Comment:
If I don't use `..` when I add new fields I need to explicitly decide if I need to use them in this pattern or if I do not need them.

But when I use `..` the compiler decides for me that I don't need to use them.

But in cases like this the only field I do not need to use are the two I explicitly do not use.

So I generally try to avoid `..` as it makes it easier for bugs to sneak in when adding new fields.

Reply:
With this reasoning we should not access fields using the dot notation because the compiler will not tell us that we have added a new one.

Reply:
Through I'm not "pattern destructing" in that case.

And if I have a use case which consumes `self` by splitting a struct into it's parts I indeed nearly always use `let` based "pattern destruction" instead of a moving out `.field` access.

Reply:
But in general the fact that not using `..` will allow us to avoid bugs does not seems too effective to me. This could be applied only when we are deconstructing a struct where we have the visibility on all the fields, while we miss all other cases (usually the majority) where the compiler cannot warn us. In my opinion this applied in general only add noise to the code because you have fields that you don't use explicitly listed and hinder maintainability because you have to change code that you would not have touched otherwise.